subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/dsl/dependencies/GradlePluginVariantsSupportTest.groovy

```diff
@@ -0,0 +1,129 @@
+
+
+import org.gradle.api.attributes.plugin.GradlePluginApiVersion
+import org.gradle.api.internal.attributes.AttributeContainerInternal
+import org.gradle.api.internal.attributes.DefaultAttributesSchema
+import org.gradle.api.internal.attributes.EmptySchema
+import org.gradle.internal.component.model.AttributeMatchingExplanationBuilder
+import org.gradle.internal.component.model.AttributeSelectionSchema
+import org.gradle.internal.component.model.ComponentAttributeMatcher
+import org.gradle.internal.isolation.IsolatableFactory
+import org.gradle.util.AttributeTestUtil
+import org.gradle.util.SnapshotTestUtil
+import org.gradle.util.TestUtil
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class GradlePluginVariantsSupportTest extends Specification {
+
+    def attributes = AttributeTestUtil.attributesFactory()
+    def objects = TestUtil.objectFactory()
+    def matcher = new ComponentAttributeMatcher()
+    def schema = new DefaultAttributesSchema(matcher, TestUtil.instantiatorFactory(), SnapshotTestUtil.valueSnapshotter() as IsolatableFactory)
+    def ep = Stub(AttributeMatchingExplanationBuilder)
+
+    def setup() {
+        GradlePluginVariantsSupport.configureSchema(schema)
+    }
+
+    @Unroll
+    def "Gradle #currentGradleVersion #accetpsOrRejects 7.0 api"() {
+        given:
+        def accepts = accetpsOrRejects == 'accepts'
+
+        when:
+        def consumer = versionAttribute(currentGradleVersion)
+        def producer = versionAttribute('7.0')
+
+        then:
+        accepts == (matcher.match(schema(), [producer], consumer, null, ep) == [producer])
+        accepts == matcher.isMatching(schema(), producer, consumer)
+
+        where:
+        currentGradleVersion       | accetpsOrRejects
+        '7.0'                      | 'accepts'
+        '7.0-20210211230048+0000'  | 'accepts'
+        '7.0-milestone-1'          | 'accepts'
+        '7.0-rc-2'                 | 'accepts'
+        '7.2'                      | 'accepts'
+        '8.0'                      | 'accepts'
+
+        '6.8.3'                    | 'rejects'
+        '6.8-rc-2'                 | 'rejects'
+        '5.0'                      | 'rejects'
+    }
+
+    def "chooses exact match API if available"() {
+        when:
+        def consumer = versionAttribute('7.0')
+        def producer = [
+            versionAttribute('6.0'),
+            versionAttribute('7.0'),
+            versionAttribute('7.0-rc-1'), // this is bad practice: targeting a not-GA version
+            versionAttribute('7.1'),
+            versionAttribute('8.0')
+        ]
+
+        then:
+        matcher.match(schema(), producer, consumer, null, ep) == [versionAttribute('7.0')]
+
+    }
+
+    def "chooses closest API"() {
+        when:
+        def consumer = versionAttribute('7.2')
+        def producer = [
+            versionAttribute('6.0'),
+            versionAttribute('7.0'),
+            versionAttribute('7.1'), // 7.1 is the closest to 7.2 in this list
```

Comment:
No. Larger versions are never selected because they are not compatible. If you have a variant for 7.3 it means it won't work with 7.2.

Maybe I should write _closest lower version_ here to clarify.